### PR TITLE
Mjcbranch421

### PR DESCRIPTION
**Pull request by michaelconnor468**

### CODE REVIEW

File: src/foam/nanos/rope/test/TestObjects.js [hunk 1]
```diff
@@ -67,63 +67,66 @@ foam.CLASS({
...
+  forwardName: 'entities',
+  inverseName: 'agents',
+  junctionDAOKey: 'ropeAgentJunctionDAO'
+});
+
+foam.RELATIONSHIP({
+  cardinality: '*:*',
+  sourceModel: 'foam.nanos.rope.test.ROPEUser',
+  targetModel: 'foam.nanos.rope.test.ROPEUser',
+  forwardName: 'parteners',
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/TestObjects.js]:
```suggestion
  forwardName: 'partners',
```

File: src/foam/nanos/rope/test/TestObjects.js [hunk 2]
```diff
@@ -67,63 +67,66 @@ foam.CLASS({
...
+  inverseName: 'sourceAccount',
+  junctionDAOKey: 'ropeTransactionDAO'
+});
+
+foam.RELATIONSHIP({
+  cardinality: '1:*',
+  sourceModel: 'foam.nanos.rope.test.ROPEBankAccount',
+  targetModel: 'foam.nanos.rope.test.ROPETransaction',
+  forwardName: 'credits',
+  inverseName: 'sourceAccount',
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/TestObjects.js]:
```suggestion
  inverseName: 'destinationAccount',
```

File: src/foam/nanos/rope/test/TestObjects.js [hunk 3]
```diff
@@ -67,63 +67,66 @@ foam.CLASS({
...
+  junctionDAOKey: 'ropeTransactionDAO'
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/TestObjects.js]:
```suggestion
  targetDAOKey: 'ropeTransactionDAO'
```

File: src/foam/nanos/rope/test/TestObjects.js [hunk 4]
```diff
@@ -67,63 +67,66 @@ foam.CLASS({
...
+  junctionDAOKey: 'ropeAccountDAO'
+});
+
+foam.RELATIONSHIP({
+  cardinality: '1:*',
+  sourceModel: 'foam.nanos.rope.test.ROPEBankAccount',
+  targetModel: 'foam.nanos.rope.test.ROPETransaction',
+  forwardName: 'debits',
+  inverseName: 'sourceAccount',
+  junctionDAOKey: 'ropeTransactionDAO'
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/TestObjects.js]:
```suggestion
  targetDAOKey: 'ropeTransactionDAO'
```

File: src/foam/nanos/rope/test/TestObjects.js [hunk 5]
```diff
@@ -67,63 +67,66 @@ foam.CLASS({
...
+  junctionDAOKey: 'ropeSigningOfficerJunctionDAO'
+});
+
+foam.RELATIONSHIP({
+  cardinality: '1:*',
+  sourceModel: 'foam.nanos.rope.test.ROPEUser',
+  targetModel: 'foam.nanos.rope.test.ROPEBankAccount',
+  forwardName: 'bankaccounts',
+  inverseName: 'owner',
+  junctionDAOKey: 'ropeAccountDAO'
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/TestObjects.js]:
```suggestion
  targetDAOKey: 'ropeAccountDAO'
```

File: src/foam/nanos/rope/test/TestObjects.js [hunk 6]
```diff
@@ -67,63 +67,66 @@ foam.CLASS({
...
+});
+
+foam.RELATIONSHIP({
+  cardinality: '1:*',
+  sourceModel: 'foam.nanos.rope.test.ROPEBankAccount',
+  targetModel: 'foam.nanos.rope.test.ROPETransaction',
+  forwardName: 'credits',
+  inverseName: 'sourceAccount',
+  junctionDAOKey: 'ropeTransactionDAO'
+});
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/TestObjects.js:131]: plz add the junction models generated from the *:* relationships to foam2/tools/classes.js under the ROPE section

File: src/foam/nanos/rope/ROPEAuthorizer.java [hunk 7]
```diff
@@ -86,7 +91,18 @@ public void authorizeOnDelete(X x, FObject obj) throws AuthorizationException {
         .getArray();
   }
 
-  public boolean ropeSearch(ROPEActions operation, FObject obj, X x, String targetDAOKey) {
+  public List<FObject> getJunctionObjects(X x, ROPE rope, FObject obj) {
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/ROPEAuthorizer.java:94]: I don't think we need this

**Review by noodlemoodle — CHANGES_REQUESTED** [review]

File: src/foam/nanos/rope/test/services.jrl [hunk 8]
```diff
@@ -87,3 +92,146 @@ p({
...
+      .setJournalType(foam.dao.JournalType.SINGLE_JOURNAL)
+      .setJournalName("ropeContacts")
+      .setAuthorize(false)
+      .setPermissioned(false)
+      .build();
+  """,
+  "client":
+  """
+    {
+      "of": "foam.nanos.rope.test.ROPEUser"
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/services.jrl]:
```suggestion
      "of": "foam.nanos.rope.test.ROPEUserROPEUserJunction"
```

File: src/foam/nanos/rope/test/services.jrl [hunk 9]
```diff
@@ -87,3 +92,146 @@ p({
...
+
+p{(
+  "class":"foam.nanos.boot.NSpec",
+  "name":"ropeContactDAO",
+  "serve":true,
+  "lazy":false,
+  "serviceScript":
+  """
+    return new foam.dao.EasyDAO.Builder(x)
+      .setOf(foam.nanos.rope.test.ROPEUser.getOwnClassInfo())
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/services.jrl]:
```suggestion
      .setOf(foam.nanos.rope.test. ROPEUserROPEUserJunction.getOwnClassInfo())
```

File: src/foam/nanos/rope/test/TestObjects.js [hunk 10]
```diff
@@ -67,63 +67,66 @@ foam.CLASS({
...
+  cardinality: '*:*',
+  sourceModel: 'foam.nanos.rope.test.ROPEUser',
+  targetModel: 'foam.nanos.rope.test.ROPEUser',
+  forwardName: 'parteners',
+  inverseName: 'partnered',
+  junctionDAOKey: 'ropePartnerJunctionDAO'
+});
+
+foam.RELATIONSHIP({
+  cardinality: '1:*',
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/TestObjects.js]:
```suggestion
  cardinality: '*:*',
```

put as *:* for now, easier to work with

**Review by noodlemoodle — CHANGES_REQUESTED** [review]

File: src/foam/nanos/rope/test/services.jrl [hunk 11]
```diff
@@ -87,3 +92,146 @@ p({
...
+
+p{(
+  "class":"foam.nanos.boot.NSpec",
+  "name":"ropeSigningOfficerJunctionDAO",
+  "serve":true,
+  "lazy":false,
+  "serviceScript":
+  """
+    return new foam.dao.EasyDAO.Builder(x)
+      .setOf(foam.nanos.rope.test.ROPEUserROPEUserJunction.getOwnClassInfo())
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/services.jrl]:
```suggestion
      .setOf(foam.nanos.rope.test.ROPEBusinessROPEUserJunction.getOwnClassInfo())
```

File: src/foam/nanos/rope/test/services.jrl [hunk 12]
```diff
@@ -87,3 +92,146 @@ p({
...
+      .setJournalType(foam.dao.JournalType.SINGLE_JOURNAL)
+      .setJournalName("ropeSiginingOfficerJunctions")
+      .setAuthorize(false)
+      .setPermissioned(false)
+      .build();
+  """,
+  "client":
+  """
+    {
+      "of": "foam.nanos.rope.test.ROPEUserROPEUserJunction"
```
Comment (noodlemoodle) [inline src/foam/nanos/rope/test/services.jrl]:
```suggestion
      "of": "foam.nanos.rope.test.ROPEBusinessROPEUserJunction"
```

**Review by noodlemoodle — CHANGES_REQUESTED** [review]

File: src/foam/nanos/rope/ROPEAuthorizer.java (hunk 7 above)
Comment (michaelconnor468) [reply src/foam/nanos/rope/ROPEAuthorizer.java:94]: it makes the code nicer and more robust

Comment (noodlemoodle) [reply src/foam/nanos/rope/ROPEAuthorizer.java:94]: how

Comment (noodlemoodle) [reply src/foam/nanos/rope/ROPEAuthorizer.java:94]: ?

**Comment by noodlemoodle** [comment]
pls revert @michaelconnor468
Merge https://github.com/michaelconnor468/foam2 into mjcbranch421
5d6eb9e

**Pull request closed without merging**